network/ClientNetworking.cpp

```diff
@@ -127,13 +128,23 @@ ClientNetworking::ClientNetworking() :
     m_io_service(),
     m_socket(m_io_service),
     m_incoming_messages(m_mutex),
-    m_connected(false),
-    m_cancel_retries(false)
+    m_rx_connected(false),
+    m_tx_connected(false)
 {}
 
-bool ClientNetworking::Connected() const {
+bool ClientNetworking::IsConnected() const {
     boost::mutex::scoped_lock lock(m_mutex);
-    return m_connected;
+    return m_rx_connected && m_tx_connected;
+}
+
+bool ClientNetworking::IsRxConnected() const {
+    boost::mutex::scoped_lock lock(m_mutex);
```

Comment:
> The processor provides special primitives that are atomic just for implementing mutex, semaphors, barries etc.

Oh, I see. Thanks for the explanation. Things definitely have progressed since I've been learning all that stuff...

> Can either of you suggest a better comment for m_mutex so that it is clear to more people, why the mutex is being used?

Maybe add that those read and write operations aren't atomic, which is why shared access has to be protected? Something along these lines:
```
// m_mutex guards m_incoming_message, m_rx_connected and m_tx_connected which are written by
// the networking thread and read by the main thread to check incoming messages and connection
// status. As those read and write operations are not atomic, shared access has to be
// protected to prevent unpredictable results.
```